source/language/subroutines.rst

```diff
@@ -1,23 +1,37 @@
 Subroutines
 ===========
 
-Subroutines are declared using the statement ``def name(parameters) -> output_type { body }``.
-The subroutine will define zero or more parameters as input, consisting of both quantum and classical arguments.
-Quantum bits and registers are passed in by reference or name, while classical types are passed in by value.
-All arguments are declared together with their type,
-for example ``qubit ancilla`` would define a quantum bit argument named ``ancilla``.
+Subroutines are declared using the statement
+
+    ``def name(parameters) -> output_type { body }``
+
+Subroutines and their named arguments must be named according to the rules for
+identifiers (See :ref:`identifiers`).
+
+The subroutine will define zero or more parameters as input, consisting of both
+quantum and classical arguments. Quantum bits and classical values are passed to
+the subroutine by reference or name in, while classical types are passed in by value.
```

Comment:
> How are classical values (e.g., 0.707+0.707im) passed by reference?

And to pour some gasoline: What is the definition of a _reference_ in OpenQASM3?

Reply:
> Unneeded "in".

Fixed, thank you.
 
> > _classical values_ are passed to the subroutine by _reference_
> 
> How are classical values (e.g., `0.707+0.707im`) passed by reference?

I have no idea, I didn't write that!

Reply:
How about this?

> Quantum [and classical*] bits are passed by reference, while classical types are passed by value.

The quantum bits truly are references because the qubit arrays and physical notation (e.g., $0) are truly just referring to a mapping in the device. The classical types being passed by value means that classical variables can't be modified by a function, indicating that the device itself incapable of performing classical numeric operations on program variables.

* Classical bits, similarly, are references to hardware allocations (not memory locations, but registers) allowing the outputs of measurements to be stored - the code is controlling the device operation, not the values. This does have the weird side-effect of potentially allowing for some devices which can perform classical operations in the control logic such as loading or storing values into the classical bits, or performing simple logical or arithmetic operations (+, -, AND, OR, etc.) like an actually assembly language. I believe that this is explicitly TSC were trying to avoid with having pass by reference variables, but I'm not sure. Also, it leads us back into memory mapping/management territory that we may want to avoid.